File: Striver/Day 26 : DP-2/Palindromic Partitioning.cpp
//Question: https://practice.geeksforgeeks.org/problems/palindromic-patitioning4845/1  

//Solution: 
int t[501][501];
    bool isPalindrome(string s,int i,int j){
        while(i<j){
            if(s[i++] != s[j--])
                return false;
        }
        return true;
    }
    int solve(string s, int i, int j){
            
        if (t[i][j] != -1)
            return t[i][j];
        if (i>=j){
            return 0;
        }
        if (isPalindrome(s,i,j))
            return 0;
            
        int ans = INT_MAX;
        for (int k=i; k<j; k++){
            int left,right;
            if(t[i][k] != -1)
                left = t[i][k];
            else t[i][k] = left = solve(s,i,k);
            
            if(t[k+1][j] != -1)
                right = t[k+1][j];
            else t[k+1][j] = right = solve(s,k+1,j);
            
            int temp = left + right + 1;
            ans = min(ans,temp);
        }
        return t[i][j] = ans;
    }
    
    int palindromicPartition(string str)
    {
        memset(t,-1,sizeof(t));
        return solve(str,0,str.length()-1);
    }

File: Striver/Day 26 : DP-2/Palindromic Partitioning.java
class Solution{
    static int palindromicPartition(String str)
    {
        int[][] dp = new int[str.length()][str.length()];
        return helper(str , 0 , str.length()-1 , dp);
    }
    
    public static int helper(String str, int i , int j , int[][] dp)
    {
        if(i>=j || isPalindrome(str , i , j) )
        return 0;
        
        if(dp[i][j] > 0)
        return dp[i][j];
        
        int ans = Integer.MAX_VALUE;
        
        for(int k=i ; k<=j-1 ; k++)
        {
            int temp_ans  =  helper(str , i , k , dp ) + helper(str , k+1 , j , dp) ;
            ans = Math.min(temp_ans +1 , ans);
        }
        
        return dp[i][j] = ans;
    }
    
    public static boolean isPalindrome(String str , int i , int j)
    {
        while(i<j)
        {
            if(str.charAt(i) == str.charAt(j))
            {
                i++;
                j--;
            }
            
            else
            return false;
        }
        
        return true;
    }
    
    
}
